Add ability to customise the pre-filled Image title (image uploads)

lb- opened the pull request:
Resolves #4945 & is a smaller scoped version of #5616 & #6696 that only covers images (not documents).

## Overview of implementation
- existing behaviour for title generation will be modified to remove the file extension from the generated title
- adds prefilling of the title on the single upload image form (not directly accessible via the UI, but code is still maintained)
- it will be easy to remove this default behaviour or add any custom title generation logic leveraging a custom DOM event provides the ability to update the title before being added to the form
- hopefully provides a minimal change but also very powerful extensibility (multiple listeners can be added and can easily access the `form` element to do read data about the collection/page or anything to inject different logic)
- does rely on mutation of data inside the event handler, but this is not really any different to what happens in Python land and the scope of any footgun is minimal, also unhandled errors will fail async so even bad JS here will not break the upload functionality (although errors are a bit harder to parse).
- Not IE11 compatible https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent - however there is already existing usage of `CustomEvent` in our code and we dropped support in [2.14](https://docs.wagtail.io/en/stable/editor_manual/browser_issues.html#ie11).

## Event Naming Convention
* Originally I had `wagtailImagesUpload`, however upon further research - all lower case is recommended.
* Options: 
  * colon for library & kebab-cas `wagtail:images-upload` (Vue recommends this)
  * dot `wagtail.images.upload` (Bootstrap uses this style)
  * all kebab `wagtail-images-upload` (other articles recommend this)
* I have gone with `wagtail:images:upload` as this will make it easier to find in code compared to other JS/HTML items.
* If we go with this we may want to consider adding other event triggers in the future with a similar convention.
* Links
  * https://eslint.vuejs.org/rules/custom-event-name-casing.html
  * https://eslint.vuejs.org/rules/custom-event-name-casing.html
  * https://stackoverflow.com/questions/19071579/javascript-dom-event-name-convention
  * https://www.w3.org/TR/1998/REC-xml-19980210#NT-Name

## Forms updated
- view - admin/images/multiple/add
- view - admin/images/add
- modal - image-chooser-modal

## How to validate
* For each of the forms above, attempt to upload an image and check the title field gets pre-filled with the filename but not the extension.
* For the non-multiple forms, before clicking save, select a different file to upload and confirm the title is left unchanged.

## PR Checklist
* Do the tests still pass? ✅ 
* Does the code comply with the style guide? ✅ 
* For Python changes: Have you added tests to cover the new/fixed behaviour? ✅ 
* For front-end changes: Did you test on all of [Wagtail’s supported environments](https://docs.wagtail.io/en/latest/contributing/developing.html#browser-and-device-support)?
    * Firefox 92 macOS
    * Chrome 95 macOS
    * Safari 14 macOS
* For new features: Has the documentation been updated accordingly? 👍

zerolab commented on wagtail/images/templates/wagtailimages/images/add.html:
```diff
@@ -11,6 +11,30 @@
+                    if ($titleField.val()) return;
+
+                    // file widget value example: `C:\fakepath\image.jpg` - convert to just the filename part
+                    var data = { title: $(this).val().split('\\').slice(-1)[0] };
```
Should this not happen in `add-multiple.js` as well for consistency?

zerolab commented:
@lb- I've tested this locally and +1 for the changes from me. Would be good someone else to have a look at the JS (multiple sets of eyes and all that)

> existing behaviour for title generation will be kept, but it will be easy to add a customisation of this behaviour

I find myself routinely dropping the extension so my preference would be to normalise the title (`file-name.ext` -> `file name`), but this may well be a biased preference.

zerolab commented on docs/advanced_topics/images/title_generation_on_upload.md:
```diff
@@ -0,0 +1,105 @@
+
+When a file is dropped into the multi-upload page or selected on the single file selection form a title will be automatically populated into the Title field for that image.
+
+Using a Javascript [event listener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) which will listen to the `'wagtailImagesUpload'` event you can customise the resolved value of this title.
```
nitpick: `s/Javascript/JavaScript/g`

lb- commented:
@zerolab  can you take another look, I have updated the items noted + a few other refinements

* updated the event name trigger, based on some research using `camelCase` is not ideal
* updated the default behaviour to remove the file extension, preventing default will stop that (multiple uploads) and stop pre-filling at all (single upload)
* updated the examples & refined the docs

zerolab reviewed (changes requested):
Hey @lb- this is looking so much better, thank you!

A few suggestions around docs and a couple of questions.

lb- commented on docs/advanced_topics/images/title_generation_on_upload.md:
```diff
@@ -0,0 +1,107 @@
+
+## DOM Event
+
+The event name to listen for is `'wagtail:images-upload'` and it will be dispatched on the image upload `form`. The event's `detail` will contain a `data` object containing the `title` that will be used (filename without extension). `detail` will also include a `maxTitleLength` integer (or `null`) which will be the max character length on the Image model's title field and `filename`, the original ...
```
To allow mutation of the title it must be in a nested object, so hence the `data` containing `title`.

These other two parts are read only, hence directly on the `detail` object.
The `maxTitleLength` is there in case customisation wants to do something smart with reducing the title length.
The `filename` is there to provide access to the raw (with file extension) name.

lb- commented on docs/advanced_topics/images/title_generation_on_upload.md:
```diff
@@ -0,0 +1,107 @@
+
+The event name to listen for is `'wagtail:images-upload'` and it will be dispatched on the image upload `form`. The event's `detail` will contain a `data` object containing the `title` that will be used (filename without extension). `detail` will also include a `maxTitleLength` integer (or `null`) which will be the max character length on the Image model's title field and `filename`, the original ...
+
+To modify the generated Image title, access and update `event.detail.data.title`, no return value is needed.
```
Good call, fixed.

lb- commented:
@zerolab  thanks again, it is great to get the feedback on the docs, the bullet point layout for the event detail is much easier to read.

Have applied changes, rebased, fixed some linting & pushed up again.

zerolab reviewed (approved):
:shipit: 🌟

lb- commented:
Discussed at core team, happy to have this approach for JS custom injection of behaviour.
